File: sokoban-python/sokoban.py
"""
This file contains the program that receives an initial state
of the Sokoban game, produces successors for each state and yields
the final result. It uses uninformed searching algorithms to search
for the answer in the search space.
"""

class SOKOBAN:
    def __init__(self, input_file):

        file = open(input_file, 'r')
        n, m = [eval(x) for x in file.readline().split()]

        self.game_map = []
        self.player = (-1, -1)
        self.box = (-1, -1)
        self.storage = (-1, -1)

        for row in range(n):
            tmp = []
            line = file.readline()
            for column in range(m):
                if line[column] == '#':
                    tmp.append(0)
                if line[column] == '.':
                    tmp.append(1)
                if line[column] == 'S':
                    self.player = (column, row)
                    tmp.append(1)
                if line[column] == '@':
                    self.box = (column, row)
                    tmp.append(1)
                if line[column] == 'X':
                    self.storage = (column, row)
                    tmp.append(1)
            self.game_map.append(tmp)
    
    def is_goal(self, box):
        # box is a tuple containing the box's coordinates
        if box == self.storage:
            return True
        return False

    def successor(self, current_state):
        current_plyr_x = current_state[0][0]
        current_plyr_y = current_state[0][1]
        current_box_x = current_state[1][0]
        current_box_y = current_state[1][1]

        # Move the player and/or the box in either of the four directions available
        # Each state is a tuple of two tuples: (Player position (x, y), Box position (x, y))
        next_states = []
        for i in [-1, 1]:
            # Move along the X axis
            next_plyr_x = current_plyr_x + i
            next_plyr_y = current_plyr_y

            if self.game_map[next_plyr_y][next_plyr_x] == 1:
                if (current_box_x == next_plyr_x and current_box_y == next_plyr_y):
                    if self.game_map[current_box_y][current_box_x + i] == 1:
                        # Move is valid
                        next_states.append(((next_plyr_x, next_plyr_y), (current_box_x + i, current_box_y)))
                else:
                    next_states.append(((next_plyr_x, next_plyr_y), (current_box_x, current_box_y)))
            
            # Move along the Y axis
            next_plyr_x = current_plyr_x
            next_plyr_y = current_plyr_y + i

            if self.game_map[next_plyr_y][next_plyr_x] == 1:
                if (current_box_x == next_plyr_x and current_box_y == next_plyr_y):
                    if self.game_map[current_box_y + i][current_box_x] == 1:
                        # Move is valid
                        next_states.append(((next_plyr_x, next_plyr_y), (current_box_x, current_box_y + i)))
                else:
                    next_states.append(((next_plyr_x, next_plyr_y), (current_box_x, current_box_y)))

        return next_states

    def test(self):
        return self.storage
    
    def initial_state(self):
        return (self.player, self.box)
    
    def bfs(self):
        init_state = self.initial_state()
        queue = []
        queue.insert(0, init_state)
        while len(queue) != 0:
            state = queue.pop()
            if self.is_goal(state[1]):
                return state
            next_steps = self.successor(state)
            for x in next_steps:
                queue.insert(0, x)

        # no goal was found
        return ((-1, -1), (-1, -1))

    def bfs_with_path(self):
        init_state = self.initial_state()
        queue = []
        queue.insert(0, [init_state])
        while len(queue) != 0:
            path = queue.pop()
            state = path[-1]
            if self.is_goal(state[1]):
                return path
            next_steps = self.successor(state)
            for x in next_steps:
                new_path = list(path)
                new_path.append(x)
                queue.insert(0, new_path)

        # no goal was found
        return ((-1, -1), (-1, -1))

File: Sokoban-cpp/sokoban.cpp
// #include <sokoban-cpp\sokoban.h>
// #include <C:\Users\InVaderrr\Documents\repos\Sokoban-game\sokoban-cpp\sokoban.h>


// SOKOBAN::SOKOBAN(const char* input_file)
//         {
//             std::ifstream infile(input_file);
            
//             infile >> n >> m;

//             for (int row = 0; row < n; row++)
//             {
//                 bool* tmp = new bool(m);
//                 char* line;
//                 infile >> line;
//                 for (int col = 0; col < m; col++)
//                 {
//                     if (line[col] == '#')
//                         tmp[col] = 0;
//                     else if (line[col] == '.')
//                         tmp[col] = 1;
//                     else if (line[col] == 'S')
//                     {
//                         tmp[col] = 1;
//                         player = std::make_pair(col, row);
//                     }
//                     else if (line[col] == '@')
//                     {
//                         tmp[col] = 1;
//                         box = std::make_pair(col, row);
//                     }
//                     else if (line[col] == 'X')
//                     {
//                         tmp[col] = 1;
//                         storage = std::make_pair(col, row);
//                     }
//                 }
//                 game_map[row] = tmp;
//             }
//         };

// bool SOKOBAN::is_goal(STATE state)
// {
//     if (state.second == this->storage)
//         return true;
//     return false;
// }

// std::vector<STATE> SOKOBAN::successor(STATE current_state)
// {
//     std::vector<STATE> next_states;
//     for (int i = -1; i < 2; i += 2)
//     {
//         // Move player one step to the left/right
//         unsigned char next_plyr_x = current_state.first.first + i;
//         unsigned char next_plyr_y = current_state.first.second;

//         // Check if the move is valid
//         if (game_map[next_plyr_y][next_plyr_y] == 1)
//         {
//             // Check if the player is moving into the box
//             if (current_state.second.first == next_plyr_x && current_state.second.second == next_plyr_y)
//             {
//                 unsigned char next_box_x = current_state.second.first + i;
//                 unsigned char next_box_y = current_state.second.second;
//                 // Check if the box is movable in this direction
//                 if (game_map[next_box_y][next_box_x] == 1)
//                     next_states.push_back(std::make_pair(
//                                     std::make_pair(next_plyr_x, next_plyr_y), 
//                                     std::make_pair(next_box_x, next_box_y)));
//             }
//             else
//             {
//                 next_states.push_back(std::make_pair(
//                                 std::make_pair(next_plyr_x, next_plyr_y), 
//                                 std::make_pair(current_state.second.first, current_state.second.second)));
//             }
//         }

//         // Move player one step upward/downward
//         next_plyr_x = current_state.first.first;
//         next_plyr_y = current_state.first.second + i;

//         // Check if the move is valid
//         if (game_map[next_plyr_y][next_plyr_y] == 1)
//         {
//             // Check if the player is moving into the box
//             if (current_state.second.first == next_plyr_x && current_state.second.second == next_plyr_y)
//             {
//                 unsigned char next_box_x = current_state.second.first;
//                 unsigned char next_box_y = current_state.second.second + i;
//                 // Check if the box is movable in this direction
//                 if (game_map[next_box_y][next_box_x] == 1)
//                     next_states.push_back(std::make_pair(
//                                     std::make_pair(next_plyr_x, next_plyr_y), 
//                                     std::make_pair(next_box_x, next_box_y)));
//             }
//             else
//             {
//                 next_states.push_back(std::make_pair(
//                                 std::make_pair(next_plyr_x, next_plyr_y), 
//                                 std::make_pair(current_state.second.first, current_state.second.second)));
//             }
            
//         }
//     }

//     return next_states;
// }

// STATE SOKOBAN::bfs()
// {
//     STATE initial_state = std::make_pair(player, box);
//     std::queue<STATE> queue;
//     queue.push(initial_state);

//     while (!queue.empty())
//     {
//         STATE state = queue.back();
//         queue.pop();
//         if (is_goal(state))
//             return state;
//         std::vector<STATE> successors = successor(state);
//         for (auto s = successors.begin(); s != successors.end(); s++)
//             queue.push(*s);
//     }

//     return std::make_pair(std::make_pair(-1, -1), std::make_pair(-1, -1));
// }